nixos/modules/security/acme.nix

```diff
@@ -288,8 +305,11 @@ in
                           ++ concatLists (mapAttrsToList (name: root: [ "-d" name ]) data.extraDomains)
                           ++ (if data.dnsProvider != null then [ "--dns" data.dnsProvider ] else [ "--http" "--http.webroot" data.webroot ])
                           ++ optionals (cfg.server != null || data.server != null) ["--server" (if data.server == null then cfg.server else data.server)];
-                runOpts = escapeShellArgs (globalOpts ++ [ "run" ]);
-                renewOpts = escapeShellArgs (globalOpts ++ [ "renew" "--days" (toString cfg.validMinDays) ]);
+                certOpts = optionals data.mustStaple [ "--must-staple" ];
+                runOpts = escapeShellArgs (globalOpts ++ [ "run" ] ++ certOpts);
+                renewOpts = escapeShellArgs (globalOpts ++
+                  [ "renew" "--days" (toString cfg.validMinDays) ] ++
+                  certOpts ++ data.additionalLegoRenewFlags);
```

Comment:
`globalOpts`, `certOpts`, `runOpts`, and `renewOpts` - each of which is determined by a load of module options. I worry the code is becoming too complicated to properly test against all potential regressions on future changes.

Anyone have any ideas to simplify this, all while adding the flexibility @emilazy has added here? :man_shrugging:

Reply:
I spent an unreasonable amount of time thinking about this already. There's just a lot of different ways to configure your certs (multiple places for email, implicit selection between dns and http for example). Maybe using some of the lib.optional* functions would help clean it up a bit. The fact that you need 2 separate strings for "run" and "renew" is a bit annoying too.

Reply:
I don't really have good ideas here, other than for lego to support proper configuration files itself or something. I currently have to do

```nix
{
  # TODO: DNS validation
  security.acme = {
    email = "...";
    acceptTerms = true;
    certs = let cfg = {
      user = "nginx";
      group = "nginx";
      mustStaple = true;
    }; in {
      "emily.moe" = cfg // { extraDomains = { "patchouli.emily.moe" = null; }; };
      "mew.build" = cfg;
    };
  };
}
```

to share options between my certificates, so the current configuration model certainly isn't ideal.